url_serde/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+///
+/// This is useful to deserialize Url types used in structure fields or
+/// tuple members with `#[serde(deserialize_with = "url_serde::deserialize")]`.
+pub fn deserialize<T, D>(deserializer: D) -> Result<T, D::Error>
+    where D: Deserializer, De<T>: Deserialize
+{
+    De::deserialize(deserializer).map(De::into_inner)
+}
+
+
+/// A wrapper to deserialize Url types.
+///
+/// This is useful with functions such as `serde_json::from_str`.
+///
+/// Values of this type can only be obtained through
+/// the `serde::Deserialize` trait.
+#[derive(Debug)]
+pub struct De<T>(T);
+
+impl<T> De<T> where De<T>: serde::Deserialize {
+    /// Consumes this wrapper, returning the deserialized value.
+    #[inline(always)]
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+
+/// Deserializes this URL from a `serde` stream.
+impl Deserialize for De<Url> {
+    fn deserialize<D>(deserializer: D) -> Result<De<Url>, D::Error> where D: Deserializer {
+        let string_representation: String = Deserialize::deserialize(deserializer)?;
+        Url::parse(&string_representation).map(De).map_err(|err| {
+            serde::de::Error::custom(err.description())
+        })
+    }
+}
+
+
+/// A convenience wrapper to be used as a type parameter, for example when
+/// a `Vec<T>` need to be passed to serde.
+#[derive(Clone, PartialEq)]
+pub struct Serde<T>(pub T)
+    where De<T>: Deserialize, for<'a> Ser<'a, T>: Serialize;
+
+impl<T> Serde<T>
+where De<T>: Deserialize, for<'a> Ser<'a, T>: Serialize
+{
+    /// Consumes this wrapper, returning the inner value.
+    #[inline(always)]
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<T> fmt::Debug for Serde<T>
+where T: fmt::Debug, De<T>: Deserialize, for<'a> Ser<'a, T>: Serialize
+{
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        self.0.fmt(formatter)
+    }
+}
+
+impl<T> Deref for Serde<T>
+where De<T>: Deserialize, for<'a> Ser<'a, T>: Serialize
+{
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.0
+    }
+}
+
+impl<T> DerefMut for Serde<T>
+where De<T>: Deserialize, for<'a> Ser<'a, T>: Serialize
+{
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.0
+    }
+}
+
+impl<T: PartialEq> PartialEq<T> for Serde<T>
+where De<T>: Deserialize, for<'a> Ser<'a, T>: Serialize
+{
+    fn eq(&self, other: &T) -> bool {
+        self.0 == *other
+    }
+}
+
+impl<T> Deserialize for Serde<T>
+where De<T>: Deserialize, for<'a> Ser<'a, T>: Serialize
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where D: Deserializer
+    {
+        De::deserialize(deserializer).map(De::into_inner).map(Serde)
+    }
+}
+
+impl<T> Serialize for Serde<T>
+where De<T>: Deserialize, for<'a> Ser<'a, T>: Serialize
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where S: Serializer
+    {
+        Ser(&self.0).serialize(serializer)
+    }
+}
+
+
+#[test]
```

Comment:
Please add a `(cd url_serde && cargo test)` line after `(cd idna && cargo test)` in `Makefile`, so that this test is run on Travis-CI. (Note that the indentation needs to be an actual tab character, not spaces.)